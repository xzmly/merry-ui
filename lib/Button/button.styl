@import '../helpers/stylus.helper.styl';default-style(color)  color #fff  border none  background color  &[data-merry-ui-outline="true"]    color color    background #fff    border: 1.8px solid color.{class-prefix}-btn  white-space nowrap  border-radius 2px  box-shadow 0 2px 5px 0 rgba(0, 0, 0, .14), 0 2px 10px 0 rgba(0, 0, 0, .1)  display inline-block  text-align center  line-height 1.5  margin 8px 8px 8px 0  padding 0 15px  font-size 14px  min-width 80px  height 32px  overflow hidden  cursor pointer  transition all .2s  &:hover    box-shadow 0 2px 5px 0 rgba(0, 0, 0, .24), 0 2px 10px 0 rgba(0, 0, 0, .24)  &:focus    outline none  &:active    opacity 0.85  //disabled样式  &[disabled]    cursor not-allowed    box-shadow 0 2px 5px 0 rgba(0, 0, 0, .16), 0 2px 10px 0 rgba(0, 0, 0, .12)    opacity 0.6  &:active    opacity none//颜色样式.{class-prefix}-btn-primary  default-style(primary-color).{class-prefix}-btn-success  default-style(success-color).{class-prefix}-btn-danger  default-style(danger-color).{class-prefix}-btn-info  default-style(info-color).{class-prefix}-btn-warning  default-style(warning-color).{class-prefix}-btn-dark  default-style(dark-color)//大小样式.{class-prefix}-btn-small  height 24px  min-width 50px  padding 0 7px  font-size 12px.{class-prefix}-btn-big  height 40px  min-width 100px