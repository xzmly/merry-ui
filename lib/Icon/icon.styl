.merry-ui-icon  width: 10em; height: 10em;  vertical-align: -0.15em;  fill: currentColor;  overflow: hidden;