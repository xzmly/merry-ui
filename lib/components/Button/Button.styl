@import '../../helpers/stylus.helper.styl';small-height = 24pxsmall-min-width = 50pxdefault-height = 32pxdefault-min-width = 80pxbig-height = 48pxbig-min-width = 110pxprefix = class-prefix + "-btn"common-style(color)  color #fff  border none  background color  &.{prefix}-outline    color color    background #fff    border: 1.2px solid colorrounded-style(height)  &.{prefix}-rounded    border-radius (height/2)px// circle 样式circle-style(height)  &.{prefix}-circle    max-width height    min-width height    border-radius 50%    padding 0.{prefix}  white-space nowrap  border-radius 2px  box-shadow 0 2px 5px 0 rgba(0, 0, 0, .14), 0 2px 10px 0 rgba(0, 0, 0, .1)  display inline-block  text-align center  line-height 1  margin 8px 8px 8px 0  padding 0 15px  font-size 14px  min-width default-min-width  height default-height  overflow hidden  cursor pointer  transition all .2s  font-weight 400  rounded-style(default-height)  circle-style(default-height)  &:hover    box-shadow 0 2px 5px 0 rgba(0, 0, 0, .24), 0 2px 10px 0 rgba(0, 0, 0, .24)  &:focus    outline none  &:active    opacity 0.85  //disabled样式  &[disabled]    cursor not-allowed    box-shadow 0 2px 5px 0 rgba(0, 0, 0, .16), 0 2px 10px 0 rgba(0, 0, 0, .12)    opacity 0.6  &:active    opacity none//颜色样式.{prefix}-primary  common-style(primary-color).{prefix}-success  common-style(success-color).{prefix}-danger  common-style(danger-color).{prefix}-info  common-style(info-color).{prefix}-warning  common-style(warning-color).{prefix}-dark  common-style(dark-color).{prefix}-gray  common-style(gray-color)  color #666 !important  box-shadow none//大小样式.{prefix}-small  height small-height  min-width small-min-width  padding 0 7px  font-size 12px  rounded-style(small-height)  circle-style(small-height).{prefix}-big  height big-height  min-width big-min-width  rounded-style(big-height)  circle-style(big-height)// block 样式.{prefix}-block  width 100%  margin: 4px