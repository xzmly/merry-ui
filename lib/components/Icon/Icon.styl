@import '../../helpers/stylus.helper.styl';.{class-prefix}-icon  font-size min-font-size  width: 1em; height: 1em;  vertical-align: -0.15em;  fill: currentColor;  overflow: hidden;