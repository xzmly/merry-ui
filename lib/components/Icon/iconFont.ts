interface IconFont {    config: string;}const customCache = new Set<string>();const creatIconFont = (arg: IconFont): void => {    const {config} = arg;    if (        !customCache.has(config) &&        config.length    ) {        const script: HTMLElement = document.createElement('script');        script.setAttribute('src', config);        script.setAttribute('data-namespace', config);        customCache.add(config);        document.body.appendChild(script);    }};export default creatIconFont;