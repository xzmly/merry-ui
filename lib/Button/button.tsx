import * as React from 'react';
import classes from '../helpers/classes';
import './button.styl';

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
    className?: string;
    children?: React.ReactNode | string | React.ReactElement
    color?: 'primary' | 'success' | 'warning' | 'danger' | 'info' | 'dark'
    size?: 'default' | 'small' | 'big'
    outline?: boolean
    
}

const Button: React.FC<ButtonProps> =
    props => {

        const {className, children, color, size, outline, ...restProps} = props;

        const colorClass: string = 'btn-' + (color || 'primary');
        const sizeClass: string = size === 'default' || !size ?  '' : 'btn-' + (size || '');

        return (
            <button data-merry-ui-outline={outline}
                    className={
                    classes(className, 'btn', colorClass, sizeClass)}
                    {...restProps}>
                <span>{children}</span>
            </button>
        );
    };

export default Button;
