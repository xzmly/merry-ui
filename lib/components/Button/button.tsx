import * as React from 'react';
import classes from '../../helpers/classes';
import Icon from '../Icon/icon'
import './button.styl';

type IconType = {
    position?: 'left' | 'right'
    name: string
}

interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
    className?: string;
    children?: React.ReactNode | string | React.ReactElement
    color?: 'primary' | 'success' | 'warning' | 'danger' | 'info' | 'dark' | 'gray'
    size?: 'default' | 'small' | 'big'
    outline?: boolean
    shape?: 'rounded' | 'circle'
    icon?: IconType
    block?: boolean
}

const Button: React.FC<ButtonProps> =
    props => {

        const {
            className, children,
            color, size,
            outline, shape,
            icon, block, ...restProps
        } = props;

        const restClassName =
            (name:string|undefined|boolean):string => name ? `btn-${name}` : '';

        const visibleLeftAndRightIcon =
            (position: string):boolean | undefined =>
                icon && shape !== 'circle' && icon.position === position

        return (
            <button className={
                        classes(className, 'btn',
                            restClassName(color || 'primary'),
                            restClassName(size === 'default' || !size ? '' : size),
                            restClassName(shape),
                            restClassName(block && 'block'),
                            restClassName(outline && 'outline')
                        )}
                    {...restProps}>
                {
                    visibleLeftAndRightIcon('left') &&
                    icon && <Icon name={icon.name} style={{marginRight: '5px'}}/>
                }
                {
                    shape !== 'circle' ?
                        <span>{children}</span> :
                        icon && <Icon name={icon.name}/>
                }
                {
                    visibleLeftAndRightIcon('right') &&
                    icon && <Icon name={icon.name} style={{marginLeft: '5px'}}/>
                }
            </button>
        );
    };

export default Button;
