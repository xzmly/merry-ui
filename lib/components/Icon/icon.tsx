import * as React from 'react';
import classes from '../../helpers/classes';
import './icon.styl';
import '../../helpers/importIcons';
import createIconFont from './iconFont';

interface IconProps extends React.SVGAttributes<SVGElement> {
    name: string;
}

interface IconComponent<P> extends React.FC<P> {
    createIconFont: typeof createIconFont
}

const Icon: IconComponent<IconProps> =
    ({className, name, ...restProps}) => {
        return (
            <svg className={`${classes(className, 'icon')}`}
                 {...restProps} >
                <use xlinkHref={`#${name}`}/>
            </svg>
        );
    };

Icon.createIconFont = createIconFont;

export default Icon;
