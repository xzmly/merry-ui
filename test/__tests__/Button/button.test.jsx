import renderer from 'react-test-renderer'import Button from '../../../lib/Button/button'import Icon from '../../../lib/Icon/icon'import { mount } from "enzyme"import React from 'react'describe('button',()=>{  it('是一个button',()=>{    const json = renderer.create(<Button/>).toJSON()    expect(json).toMatchSnapshot()  })  it('接受一个className',()=>{    const c = mount(<Button className='btn'/>)    expect(c.find('button').hasClass('btn')).toBe(true)  })  it('接受disabled',()=>{    const c = mount(<Button className='btn' disabled/>)    expect(c.props().disabled).toBe(true)    c.setProps({ disabled: false });    expect(c.props().disabled).toBe(false)  })  it('接受children',()=>{    let c = mount(<Button className='btn'>123123</Button>)    expect(c.find('button > span').text()).toBe("123123")    c = mount(<Button className='btn'><span>123123</span></Button>)    expect(c.contains(<span>123123</span>)).toBe(true)    c = mount(<Button className='btn'><Icon/></Button>)    expect(c.find('svg').hasClass('merry-ui-icon')).toBe(true)  })  it('接收color',()=>{    const c = mount(<Button className='btn' color={'primary'}/>)    const _expect = (className,props,expected) =>{      expect(c.find('button').hasClass(className)).toBe(expected)      expect(c.props().color).toBe(props)    }    _expect('merry-ui-btn-primary','primary',true)    c.setProps({ color: 'success' });    _expect('merry-ui-btn-success','success',true)    c.setProps({ color: 'warning' });    _expect('merry-ui-btn-warning','warning',true)    c.setProps({ color: 'danger' });    _expect('merry-ui-btn-danger','danger',true)    c.setProps({ color: 'info' });    _expect('merry-ui-btn-info','info',true)    c.setProps({ color: 'dark' });    _expect('merry-ui-btn-dark','dark',true)    c.setProps({ color: 'gray' });    _expect('merry-ui-btn-gray','gray',true)  })  //'default' | 'small' | 'big'  it('接受size',()=>{    const c = mount(<Button className='btn' size={'default'}/>)    const _expect = (className,props,expected) =>{      expect(c.find('button').hasClass(className)).toBe(expected)      expect(c.props().size).toBe(props)    }    _expect('merry-ui-btn-default','default',false)    c.setProps({ color: 'small' });    _expect('merry-ui-btn-small','small',true)    c.setProps({ color: 'big' });    _expect('merry-ui-btn-big','big',true)  })})