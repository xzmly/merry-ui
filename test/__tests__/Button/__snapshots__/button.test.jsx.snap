// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button 是一个button 1`] = `
<button
  className="merry-ui-btn merry-ui-btn-primary"
>
  <span />
</button>
`;
